ui/js/controllers/logviewer.js

```diff
@@ -33,13 +32,36 @@ logViewerApp.controller('LogviewerCtrl', [
         $scope.currentLineNumber = 0;
         $scope.highestLine = 0;
         $scope.showSuccessful = true;
-
         $scope.$watch('artifact', function () {
             if (!$scope.artifact) {
                 return;
             }
             $scope.showSuccessful = !$scope.hasFailedSteps();
         });
+        $scope.$watch('[selectedBegin, selectedEnd]', function(newVal, oldVal) {
+            var newHash = (newVal[0] == newVal[1])? newVal[0] : newVal[0] + "-L"+newVal[1];
+            if (!isNaN(newVal[0])) {
+                $location.hash("L"+newHash);
+            }
+        });
+        $scope.$on("$locationChangeSuccess", function($event, $artifact) {
+            getSelectedLines();
+        });
+
+        $scope.click = function(line, $event) {
+            if($event.shiftKey) {
+                if (line.index < $scope.selectedBegin){
```

Comment:
It's sort of our aspired style in our codebase (not always consistent :) ), but could add white space here if you like:

`selectedBegin) {`